File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression_user.py
# Polynomial Regression

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values
y = dataset.iloc[:, -1].values

# Splitting the dataset into the Training set and Test set   --Not Needed, Small Dataset
"""from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)"""

# Feature Scaling
"""from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
sc_y = StandardScaler()
y_train = sc_y.fit_transform(y_train)"""

# Fitting Linear Regression to the dataset
from sklearn.linear_model import LinearRegression
linreg = LinearRegression()
linreg.fit(X, y)

# Fitting Polynomial Regression to the dataset
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree=4)
X_poly = poly_reg.fit_transform(X)
linreg2 = LinearRegression()
linreg2.fit(X_poly, y)

# Visualizing the Linear Regression results
plt.scatter(X, y, color="red")
plt.plot(X, linreg.predict(X), color="blue")
plt.title("Truth or bluff (Linear Regression)")
plt.xlabel("Position Level")
plt.ylabel("Salary")
plt.show()

# Visualizing the Polynomial Regression results
X_grid = np.arange(min(X), max(X), 0.1)
X_grid = X_grid.reshape((len(X_grid), 1))
plt.scatter(X, y, color="red")
plt.plot(X_grid, linreg2.predict(poly_reg.fit_transform(X_grid)), color="blue")
plt.title("Truth or bluff (Polynomial Regression)")
plt.xlabel("Position Level")
plt.ylabel("Salary")
plt.show()

# Predicting a new result with Linear Regression
linreg.predict(np.asarray(6.5).reshape(1,-1))

# Predicting a new result with Linear Regression
linreg2.predict(poly_reg.fit_transform(np.asarray(6.5).reshape(1,-1)))

File: Part 2 - Regression/Section 6 - Polynomial Regression/polynomial_regression_user.R
# Polynomial Regression

# Importing the dataset
dataset = read.csv('Position_Salaries.csv')
dataset = dataset[2:3]

# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
# library(caTools)
# set.seed(123)
# split = sample.split(dataset$DependentVariable, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# test_set = subset(dataset, split == FALSE)

# Feature Scaling
# training_set = scale(training_set)
# test_set = scale(test_set)

# Fitting Linear Regression to the dataset
linreg = lm(formula = Salary ~ Level, data = dataset)
summary(linreg)


# Fitting Polynomial Regression to the dataset
dataset$Level2 = dataset$Level ^ 2
dataset$Level3 = dataset$Level ^ 3
dataset$Level4 = dataset$Level ^ 4
polyreg = lm(formula = Salary ~ ., data = dataset)
summary(polyreg)

# Visualizing the Linear Regression results
library(ggplot2)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), 
             colour = "red") +
  geom_line(aes(x = dataset$Level, y = predict(linreg, newdata = dataset)),
            colour = "blue") +
  ggtitle("Truth or bluff (Linear Regression)") +
  xlab("Position Level") +
  ylab("Salary")

# Visualizing the Polynomial Regression results
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.1)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary), 
             colour = "red") +
  geom_line(aes(x = x_grid, y = predict(polyreg, newdata = data.frame(Level = x_grid,  Level2 = x_grid^2,
                                                                                       Level3 = x_grid^3,
                                                                                       Level4 = x_grid^4))),
            colour = "blue") +
  ggtitle("Truth or bluff (Polynomial Regression)") +
  xlab("Position Level") +
  ylab("Salary")


# Predicting a new result with Linear Regression
y_pred = predict(linreg, data.frame(Level = 6.5))

# Predicting a new result with Linear Regression
y_pred_poly = predict(polyreg, data.frame(Level = 6.5, 
                                          Level2 = 6.5^2,
                                          Level3 = 6.5^3,
                                          Level4 = 6.5^4))
